crates/stef_derive/src/lib.rs

```diff
@@ -0,0 +1,380 @@
+
+                // *t = Box::new(effect_name.clone());
+
+                (rarr, actual)
+            }
+        };
+        original_func.sig.ident = syn::Ident::new(
+            &format!("_stef_impl_{}", original_func.sig.ident),
+            f.f.span(),
+        );
+        original_func.sig.output = syn::ReturnType::Type(*rarr, output_type);
+        // let impl_name = f.impl_name();
+        // let block = f.f.block;
+        // let args = delim::<_, Token!(,)>(f.inputs().iter().map(|(pat, ty)| quote! { #pat: #ty }));
+        original_func.to_token_stream()
+        // quote! {
+        //    fn #impl_name(&mut self, #args) -> #output_type {
+        //        #block
+        //    }
+        // }
+    }));
+
+    let define_public_fns_inner = ss_flatten(fns.iter().map(|f| {
+        let mut original_func = f.f.clone();
+        let variant_name = f.variant_name();
+
+        let pats = delim::<_, Token!(,)>(f.inputs().into_iter().map(|(pat, _)| pat));
+        let arg = match pats.is_empty() {
+            true => quote! { <Self as stef::State>::Action::#variant_name },
+            false => quote! { <Self as stef::State>::Action::#variant_name(#pats) },
+        };
+
+        let new_block = if let Some(MatchPat { var, pat }) = f.match_pat.as_ref() {
+            quote! {{
+                use stef::State;
+                let eff = self.transition(#arg);
+                match eff {
+                    #pat => #var,
+                    _ => unreachable!("stef::state has a bug in its effect unwrapping logic")
+                }
+            }}
+        } else {
+            quote! {{
+                use stef::State;
+                self.transition(#arg)
+            }}
+        };
+
+        let ts = proc_macro::TokenStream::from(new_block.into_token_stream());
+        original_func.block = syn::parse(ts).expect("problem 3!");
+        original_func.to_token_stream()
+    }));
+
+    let mut define_hidden_fns: syn::ItemImpl = syn::parse(
+        quote! {
+            impl #struct_path {
+                #define_hidden_fns_inner
+            }
+        }
+        .into(),
+    )
+    .expect("problem 4!");
+    define_hidden_fns.generics = item.generics.clone();
+
+    let mut define_public_fns: syn::ItemImpl = syn::parse(
+        quote! {
+            impl #struct_path {
+                #define_public_fns_inner
+            }
+        }
+        .into(),
+    )
+    .expect("problem 5!");
+    define_public_fns.generics = item.generics.clone();
+
+    // let action_name_generic = match action_name.clone() {
+    //     Type::Path(mut path) => {
+    //         path.path.segments.last_mut().expect("problem 6!").arguments = struct_path
+    //             .path
+    //             .segments
+    //             .last()
+    //             .expect("problem 7!")
+    //             .arguments
+    //             .clone();
+    //         path
+    //     }
+    //     _ => todo!(),
+    // };
```

Comment:
```suggestion
```

Reply:
I'm not ready to delete these comments yet, they represent precious hard-won knowledge of the arcane art of writing proc macros, and this macro is not in its final form